rfcs/rfcXXXX.md

```diff
@@ -0,0 +1,154 @@
+    ID:
+    Status:  Draft
+
+## Abstract
+
+As of Perl version 5.34, `overload` is incomplete and surprising for
+string operations:
+
+* the `join` builtin should be using the `concat` (`.`) overload
+  when operating on overloaded objects (including the separator),
+* `overload` should support the `substr` operation.
+
+## Motivation
+
+The perl builtin `join`, when operating on overloaded objects will
+stringify them and the result will no longer be an object of the class
+with overloads, even if the `concat` (`.`) function is overloaded.
+
+Additionally, the `overload` module does not currently support overloading
+the `substr` operation.
+
+The current behaviour is inconsistent and confusing.
+
+There is at least one CPAN module which works around the `join`
+deficiency, (the [`join` function of
+`String::Tagged`](https://metacpan.org/pod/String::Tagged#join), but not
+universally.
+
+The CPAN module
+`[overload::substr](https://metacpan.org/pod/overload::substr)`) is a
+complete solution, but we think that deferring to CPAN when Perl should
+do the right thing is not a satisfying answer.
+
+## Rationale
+
+Extending `overload` to support `substr`, and `join` to use `concat`
+would be less surprising and more consistent with the "Do What I Mean"
+spirit of Perl.
+
+Adding this to Perl would make `overload::substr` obsolete and simplify
+the implementation of `String::Tagged` and other similar modules which
+have written their own `join`.
+
+## Specification
+
+The documentation for `join` would be amended with the following:
+
+> If any of the arguments (`EXPR` or `LIST` contents) are objects which
+> overload the concat (`.`) operation, then that overloaded operation
+> will be invoked.
+
+The documentation for `overload` would be amended to extend the complete
+list of keys that can be specified in the `use overload` statement to
+include `substr`. The module `overload::substr` code and documentation
+should be used as a reference and guide for implementation in core. Note
+that the current documentation for this module highlights a need for
+additional tests, which must be a part of a core implementation.
+
+An open question is whether `split` requires an overload target,
+and should a fallback implementation be provided using the
+overloaded `substr`?
+
+## Backwards Compatibility
+
+If code relies upon the `join` function to convert objects to plain
+strings, this would break that code, thus a feature flag seems
+necessary.
+
+There is no need autogenerate an implementation for `substr` if it is
+missing, as the stringification will happen as it does now.
+
+Modules which implement their own `join` would need to add
+version-conditional logic to work with different versions of Perl.
+
+`overload::substr` would become an optional dependency for modules
+that needs to support older Perls.
+
+## Security Implications
+
+We do not yet foresee security issues. Guidance is welcome.
+
+## Examples
+
+For `join`:
+
+    ```perl
```

Comment:
Use either indentation or code fences for verbatim blocks, not both. Code fences are better, since they let you specify the language.